crates/trie-parallel/src/async_root.rs

```diff
@@ -0,0 +1,293 @@
+use tracing::*;
+
+/// Async state root calculator.
+///
+/// The calculator starts off by launching tasks to compute storage roots.
+/// Then, it immediately starts walking the state trie updating the necessary trie
+/// nodes in the process. Upon encountering a leaf node, it will poll the storage root
+/// task for the corresponding hashed address.
+///
+/// Internally, the calculator uses [ConsistentDbView] since
+/// it needs to rely on database state saying the same until
+/// the last transaction is open.
+/// See docs of using [ConsistentDbView] for caveats.
+///
+/// For sync usage, take a look at `ParallelStateRoot`.
+#[derive(Debug)]
+pub struct AsyncStateRoot<DB, Provider> {
+    /// Consistent view of the database.
+    view: ConsistentDbView<DB, Provider>,
+    /// Blocking task pool.
+    blocking_pool: BlockingTaskPool,
+    /// Changed hashed state.
+    hashed_state: HashedPostState,
+}
+
+impl<DB, Provider> AsyncStateRoot<DB, Provider> {
+    /// Create new async state root calculator.
+    pub fn new(
+        view: ConsistentDbView<DB, Provider>,
+        blocking_pool: BlockingTaskPool,
+        hashed_state: HashedPostState,
+    ) -> Self {
+        Self { view, blocking_pool, hashed_state }
+    }
+}
+
+impl<DB, Provider> AsyncStateRoot<DB, Provider>
+where
+    DB: Database + Clone + 'static,
+    Provider: DatabaseProviderFactory<DB> + Clone + Send + Sync + 'static,
+{
+    /// Calculate incremental state root asynchronously.
+    pub async fn incremental_root(self) -> Result<B256, AsyncStateRootError> {
+        self.calculate(false).await.map(|(root, _)| root)
+    }
+
+    /// Calculate incremental state root with updates asynchronously.
+    pub async fn incremental_root_with_updates(
+        self,
+    ) -> Result<(B256, TrieUpdates), AsyncStateRootError> {
+        self.calculate(true).await
+    }
+
+    async fn calculate(
+        self,
+        retain_updates: bool,
+    ) -> Result<(B256, TrieUpdates), AsyncStateRootError> {
+        let prefix_sets = self.hashed_state.construct_prefix_sets();
+        let storage_root_targets = StorageRootTargets::new(
+            self.hashed_state.accounts.keys().copied(),
+            prefix_sets.storage_prefix_sets,
+        );
+        let hashed_state_sorted = Arc::new(self.hashed_state.into_sorted());
+
+        // Pre-calculate storage roots async for accounts which were changed.
+        debug!(target: "trie::async_state_root", len = storage_root_targets.len(), "pre-calculating storage roots");
+
+        let mut storage_roots = HashMap::with_capacity(storage_root_targets.len());
+        for (hashed_address, prefix_set) in
+            storage_root_targets.into_iter().sorted_unstable_by_key(|(address, _)| *address)
+        {
+            let view = self.view.clone();
+            let hashed_state_sorted = hashed_state_sorted.clone();
+            let handle =
+                self.blocking_pool.spawn_fifo(move || -> Result<_, AsyncStateRootError> {
```

Comment:
what's more performant here, spawning jobs or using pariter?

did you try chunking?